core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+            pageBuffer.clear();
+            bufferRetainedSizeInBytes = 0;
+            unblockIfNecessary(blocked);
+            return;
+        }
+        maxBufferRetainedSizeInBytes = max(maxBufferRetainedSizeInBytes, bufferRetainedSizeInBytes);
+        pageBuffer.putAll(taskId, pages);
+    }
+
+    @Override
+    public synchronized void taskFinished(TaskId taskId)
+    {
+        if (closed) {
+            return;
+        }
+
+        checkState(allTasks.contains(taskId), "task is not registered: %s", taskId);
+        checkState(!failedTasks.containsKey(taskId), "task is failed: %s", taskId);
+        checkState(finishedTasks.add(taskId), "task is finished: %s", taskId);
+
+        if (retryPolicy == TASK) {
+            // TODO implement deduplication for task level retires
+            throw new UnsupportedOperationException("task level retry policy is unsupported");
+        }
+        checkInputFinished();
+    }
+
+    @Override
+    public synchronized void taskFailed(TaskId taskId, Throwable t)
+    {
+        if (closed) {
+            return;
+        }
+
+        checkState(allTasks.contains(taskId), "task is not registered: %s", taskId);
+        checkState(!finishedTasks.contains(taskId), "task is finished: %s", taskId);
+        checkState(failedTasks.put(taskId, t) == null, "task is already failed: %s", taskId);
+        checkInputFinished();
+    }
+
+    @Override
+    public synchronized void noMoreTasks()
+    {
+        if (closed) {
+            return;
+        }
+
+        noMoreTasks = true;
+        checkInputFinished();
+    }
+
+    private synchronized void checkInputFinished()
+    {
+        if (failure != null) {
+            return;
+        }
+
+        if (inputFinished) {
+            return;
+        }
+
+        if (!noMoreTasks) {
+            return;
+        }
+
+        if (allTasks.isEmpty()) {
+            inputFinished = true;
+            unblockIfNecessary(blocked);
+            return;
+        }
+
+        switch (retryPolicy) {
+            case TASK:
+                // TODO implement deduplication for task level retires
+                throw new UnsupportedOperationException("task level retry policy is unsupported");
+            case QUERY: {
+                int latestAttemptId = allTasks.stream()
+                        .mapToInt(TaskId::getAttemptId)
+                        .max()
+                        .orElseThrow(() -> new VerifyException("allTasks is not expected to be empty"));
+
+                Set<TaskId> latestAttemptTasks = allTasks.stream()
+                        .filter(taskId -> taskId.getAttemptId() == latestAttemptId)
+                        .collect(toImmutableSet());
+
+                if (finishedTasks.containsAll(latestAttemptTasks)) {
+                    removePagesForPreviousAttempts(latestAttemptId);
+                    inputFinished = true;
+                    unblockIfNecessary(blocked);
```

Comment:
return here?
If we reach this place it means that `failures` below will be empty, right?

Reply:
Yeah